src/m4a.c

```diff
@@ -963,243 +967,233 @@ void CgbSound(void)
         }
 
         prevC15 = soundInfo->c15;
-        evAdd = *nrx2ptr;
+        envelopeStepTimeAndDir = *nrx2ptr;
 
-        if (channels->sf & 0x80)
+        /* 2. calculate envelope volume */
+        if (channels->statusFlags & SOUND_CHANNEL_SF_START)
         {
-            if (!(channels->sf & 0x40))
+            if (!(channels->statusFlags & SOUND_CHANNEL_SF_STOP))
             {
-                channels->sf = 3;
-                channels->mo = 3;
+                channels->statusFlags = SOUND_CHANNEL_SF_ENV_ATTACK;
+                channels->modify = CGB_CHANNEL_MO_PIT | CGB_CHANNEL_MO_VOL;
                 CgbModVol(channels);
                 switch (ch)
                 {
                 case 1:
-                    *nrx0ptr = channels->sw;
+                    *nrx0ptr = channels->sweep;
                     // fallthrough
                 case 2:
-                    *nrx1ptr = ((u32)channels->wp << 6) + channels->le;
-                    goto loc_82E0E30;
+                    *nrx1ptr = ((u32)channels->wavePointer << 6) + channels->length;
+                    goto init_env_step_time_dir;
                 case 3:
-                    if ((u32)channels->wp != channels->cp)
+                    if (channels->wavePointer != channels->currentPointer)
                     {
                         *nrx0ptr = 0x40;
-                        REG_WAVE_RAM0 = channels->wp[0];
-                        REG_WAVE_RAM1 = channels->wp[1];
-                        REG_WAVE_RAM2 = channels->wp[2];
-                        REG_WAVE_RAM3 = channels->wp[3];
-                        channels->cp = (u32)channels->wp;
+                        REG_WAVE_RAM0 = channels->wavePointer[0];
+                        REG_WAVE_RAM1 = channels->wavePointer[1];
+                        REG_WAVE_RAM2 = channels->wavePointer[2];
+                        REG_WAVE_RAM3 = channels->wavePointer[3];
+                        channels->currentPointer = channels->wavePointer;
                     }
                     *nrx0ptr = 0;
-                    *nrx1ptr = channels->le;
-                    if (channels->le)
-                        channels->n4 = -64;
+                    *nrx1ptr = channels->length;
+                    if (channels->length)
+                        channels->n4 = 0xC0;
                     else
-                        channels->n4 = -128;
+                        channels->n4 = 0x80;
```

Comment:
Is there a reason for these hex value changes? Should these be constants?

Reply:
Making a define for those would indeed be nicer. I wrote it in hex since it's a bitmap written to a register and thus is easier to read.